actions/file_actions.jsx

```diff
@@ -11,68 +10,101 @@ import {Client4} from 'mattermost-redux/client';
 
-            const failure = {
-                type: FileTypes.UPLOAD_FILES_FAILURE,
-                clientIds: [clientId],
-                channelId,
-                rootId,
-                error: err,
+        if (onProgress && xhr.upload) {
+            xhr.upload.onprogress = (event) => {
+                const percent = parseInt((event.loaded / event.total) * 100, 10);
+                onProgress({
+                    clientId,
+                    name,
+                    percent,
+                    type,
+                });
             };
+        }
+
+        if (onSuccess) {
+            xhr.onload = () => {
+                if (xhr.status === 201 && xhr.readyState === 4) {
+                    const response = JSON.parse(xhr.response);
+                    const data = response.file_infos.map((fileInfo, index) => {
+                        return {
+                            ...fileInfo,
+                            clientId: response.client_ids[index],
+                        };
+                    });
+
+                    dispatch(batchActions([
+                        {
+                            type: FileTypes.RECEIVED_UPLOAD_FILES,
+                            data,
+                            channelId,
+                            rootId,
+                        },
+                        {
+                            type: FileTypes.UPLOAD_FILES_SUCCESS,
+                        },
+                    ]));
 
-            dispatch(batchActions([failure, getLogErrorAction(err)]));
-            return {error: e};
+                    onSuccess(response, channelId, rootId);
+                }
+            };
         }
-        const data = res.body.file_infos.map((fileInfo, index) => {
-            return {
-                ...fileInfo,
-                clientId: res.body.client_ids[index],
+
+        if (onError) {
+            xhr.onerror = () => {
+                if (xhr.readyState === 4 && xhr.responseText.length !== 0) {
+                    const errorResponse = JSON.parse(xhr.response);
+
+                    forceLogoutIfNecessary(errorResponse, dispatch, getState);
+
+                    const uploadFailureAction = {
+                        type: FileTypes.UPLOAD_FILES_FAILURE,
+                        clientIds: [clientId],
+                        channelId,
+                        rootId,
+                        error: errorResponse,
+                    };
+
+                    dispatch(batchActions([uploadFailureAction, getLogErrorAction(errorResponse)]));
+                    onError(errorResponse, clientId, channelId, rootId);
+                } else {
+                    const errorMessage = xhr.status === 0 || !xhr.status ? {message: Utils.localizeMessage('file_upload.generic_error', 'There was a problem uploading your files.')} : {message: Utils.localizeMessage('channel_loader.unknown_error', 'We received an unexpected status code from the server.') + ' (' + xhr.status + ')'};
```

Comment:
Nitpicking, but could you split the ternary across multiple lines to make it more readable like
```js
const errorMessage = xhr.status === 0 || !xhr.status ?
    {message: ...} :
    {message: ...};
```

Also, you could consider making errorMessage just the string and then passing `{message: errorMessage}` into `onError` below to avoid repeating a bit. 0/5 though